L62-core-call-credential-security-level.md

```diff
@@ -0,0 +1,121 @@
+a corresponding call credential. Whenever a call credential is about to be
+transferred on a channel, we check if the security level of channel is greater than or
+equal to that of call credential and fail the transfer if it is not.
+
+## Background
+
+In gRPC Java and Go, each call credential can dictate whether or not
+a secure channel (i.e., ensuring both privacy and integrity) is needed
+for its communication. The use of secure channel is mandatory if a call
+credential is used to access Google APIs. In gRPC C core and its wrapped
+languages, a call credential can only be communicated over a secure channel
+which is also acceptable from the security perspective. However with the
+introduction of local credential that can be used in both UDS and TCP loopback
+connections, any call credential (including the ones used to access Google APIs)
+can be transferred over a TCP loopback connection that is considered insecure, which
+greatly lowers the security bar. In more detals, UDS connections can be treated
+as relatively secure (assuming a correct directory permission is set on a file path)
+as a remote peer can be authenticated and data confidentiality can be guaranteed
+against passive adversaries. However for local TCP, we cannot authenticate a
+remote peer and thus both data confidentiality and integrity become meaningless.
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+- A call credential should have an ability to dictate the security level of
+  a connection over which it will be communicated. It is an inherent property that
+  should be determined by its implementation, not the application that uses it.
+  If we give applications the power of setting security level, there will be risks that
+  applications could set it incorrectly either intentionally or by mistake.
+  Furthermore, by treating the security level as an inherent property the
+  behavior will be consistent with gRPC Java and Go implementations.
+
+  The security level should be one of privacy+integrity, integrity-only, and
+  insecure, which can be ordered from the highest (privacy+integrity) to lowest
+  (insecure). In case of a composite call credential, its security level should be set
+  to the highest level of any of its individual call credentials.
+
+- A security level should be also associated with each connection created by a channel
+  credential to indicate the level of transport security that connection is granted.
+  For a local connection, its security level will be set to privacy+integrity for UDS and
+  insecure for TCP loopback.
+
+
+- Before sending a call credential over a connection, we will extract the security levels
+  of call credential and the connection and check if the security level of
+  the connection is greater than or equal to that of call credential and only allow
+  the transfer of call credential if the above check succeeds. If the connection
+  does not meet the required security level, the call credential transfer will
```

Comment:
We should be more precise here: the entire call will fail.

Reply:
The description is updated.